### PR TITLE
EntityManager: Implement QueryExecutor & Queryable

### LINKED ISSUE #217: QueryObjects: Use EntityManager
Hi,
I've created library which is basically simpler implementation of your query objects, but it works with `EntityManager` instead of `EntityRepository`.

This is all good but this morning I had an idea, what if I implement `QueryExecutor` and `Queryable` in the `EntityManager` itself? It would need some work to unite argument API of the methods.

I wanted to ask on your opinion about this. I'd be more happier to have this directly in *Kdyby/Doctrine* instead of implementing it myself.

### PR DESCRIPTION
**Pull request by juniwalk**
Implementation of QueryExecutor & Queryable interfaces on EntityManager itself to cut the need to use EntityRepositories.

Closes #217

### CODE REVIEW

**Comment by juniwalk** [comment]
I think it would be handy to have a way to get the `QueryBuilder` from the `QueryObject`. For example `DoctrineDataSource` from *Grido* needs it. Now I have to create separate query.

Something like that:
```php
/** @var \Kdyby\Doctrine\QueryBuilder $qb */
$qb = (new UserQuery)->withGroups()
    ->getQuery($entityManager);
```

//cc: @fprochazka

File: src/Kdyby/Doctrine/EntityManager.php [hunk 1]
```diff
@@ -281,6 +284,88 @@ public function onDaoCreate(EntityManager $em, Doctrine\Common\Persistence\Objec
...
+
+	/**
+	 * Create a new QueryBuilder instance that is pre-populated for this entity name
+	 *
+	 * @param  string|NULL  $alias
+	 * @param  string|NULL  $indexBy
+	 * @param  string|NULL  $entityName
+	 * @return \Kdyby\Doctrine\DqlSelection
+	 */
+	public function select($alias = NULL, $indexBy = NULL, $entityName = NULL)
```
Comment (fprochazka) [inline src/Kdyby/Doctrine/EntityManager.php]: select was more of an experiment, I wouldn't wanna have it here

Comment (fprochazka) [inline src/Kdyby/Doctrine/EntityManager.php]: also, it's not a part of the QueryExecutor interface

Comment (fprochazka) [inline src/Kdyby/Doctrine/EntityManager.php]: dammit, it's part of queryable... :-/

Comment (juniwalk) [inline src/Kdyby/Doctrine/EntityManager.php]: Yeah that is right. Having `$alias` and `$indexBy` in `EntityManager::createQueryBuilder` isnt exactly clear either but it is required by the interface.

Comment (fprochazka) [inline src/Kdyby/Doctrine/EntityManager.php]: Is it a BC Break to remove a method from the `Queryable`  interface?

Comment (juniwalk) [inline src/Kdyby/Doctrine/EntityManager.php]: To remove should not be, add would be.

Comment (juniwalk) [inline src/Kdyby/Doctrine/EntityManager.php]: Shall I add it into my PR?

Comment (fprochazka) [inline src/Kdyby/Doctrine/EntityManager.php]: Yes please.

**Comment by juniwalk** [comment]
Alright, select is gone now.

File: src/Kdyby/Doctrine/EntityManager.php [hunk 2]
```diff
@@ -281,6 +284,68 @@ public function onDaoCreate(EntityManager $em, Doctrine\Common\Persistence\Objec
...
+
+
+
+	/**
+	 * @param \Kdyby\Persistence\Query|\Kdyby\Doctrine\QueryObject $queryObject
+	 * @param int $hydrationMode
+	 * @throws QueryException
+	 * @return array|\Kdyby\Doctrine\ResultSet
+	 */
+	public function fetch(Persistence\Query $queryObject, $hydrationMode = AbstractQuery::HYDRATE_OBJECT)
```
Comment (fprochazka) [inline src/Kdyby/Doctrine/EntityManager.php]: Now let's hope doctrine won't decide to create a `fetch` method for something here :)

**Comment by fprochazka** [comment]
I like it now. Is there anything that you'd like to add?

File: src/Kdyby/Doctrine/EntityManager.php (hunk 2 above)
Comment (juniwalk) [inline src/Kdyby/Doctrine/EntityManager.php]: That would be unfortunate!

**Comment by juniwalk** [comment]
Well there is [this thing](#issuecomment-155372323), but that can be added in separate PR if we get to some solution.

I think I am done.

File: src/Kdyby/Doctrine/EntityManager.php [hunk 3]
```diff
@@ -74,7 +77,7 @@ public function bindTracyPanel(Diagnostics\Panel $panel)
 	/**
 	 * @return \Kdyby\Doctrine\QueryBuilder
 	 */
-	public function createQueryBuilder()
+	public function createQueryBuilder($alias = NULL, $indexBy = NULL)
```
Comment (enumag) [inline src/Kdyby/Doctrine/EntityManager.php]: These parameters are not used. I know why but still... :-(

Comment (juniwalk) [inline src/Kdyby/Doctrine/EntityManager.php]: I don't like it either.

I was trying to go the way of least resistance, ofcourse they could be removed from the interface but I have no idea what that would do with existing code.

Comment (fprochazka) [inline src/Kdyby/Doctrine/EntityManager.php]: That would definitely be a bc break :-/

Comment (fprochazka) [inline src/Kdyby/Doctrine/EntityManager.php]: We might consider throw an `NotImplementedException` when they're not `=== NULL`

Comment (enumag) [inline src/Kdyby/Doctrine/EntityManager.php]: @fprochazka :+1:

Comment (juniwalk) [inline src/Kdyby/Doctrine/EntityManager.php]: Alright, I will add it.

Comment (juniwalk) [inline src/Kdyby/Doctrine/EntityManager.php]: Will this be sufficient?
```php
if ($alias || $indexBy) {
    throw new NotSupportedException('Use EntityRepository for $alias and $indexBy arguments to work.');
}
```

**Comment by fprochazka** [comment]
:+1:

**Comment by fprochazka** [comment]
Thanks!

**Comment by juniwalk** [comment]
@matej21 Why not? When you use repository for queryable you can use it the same way, can you not?

**Pull request merged**